<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multivariate Genome Wide Marginal Epistasis Test • mvMAPIT</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Multivariate Genome Wide Marginal Epistasis Test">
<meta property="og:description" content="Epistasis, commonly defined as the interaction between genetic
  loci, is known to play an important role in the phenotypic variation of
  complex traits. As a result, many statistical methods have been developed to
  identify genetic variants that are involved in epistasis, and nearly all of
  these approaches carry out this task by focusing on analyzing one trait at a
  time. Previous studies have shown that jointly modeling multiple phenotypes
  can often dramatically increase statistical power for association mapping. In
  this  package, we present the multivariate MArginal ePIstasis Test
  (mvMAPIT) – a multi-outcome generalization of a recently proposed epistatic
  detection method which seeks to detect marginal epistasis or the combined
  pairwise interaction effects between a given variant and all other variants.
  By searching for marginal epistatic effects, one can identify genetic variants
  that are involved in epistasis without the need to identify the exact
  partners with which the variants interact – thus, potentially alleviating
  much of the statistical and computational burden associated with conventional
  explicit search based methods. Our proposed mvMAPIT builds upon this
  strategy by taking advantage of correlation structure between traits to
  improve the identification of variants involved in epistasis.
  We formulate mvMAPIT as a  multivariate linear mixed model and develop a
  multi-trait variance component estimation algorithm for efficient parameter
  inference and P-value computation. Together with reasonable model
  approximations, our proposed approach is scalable to moderately sized
  genome-wide association studies.
  Crawford et al. (2017) &lt;doi:10.1371/journal.pgen.1006869&gt;.
  Stamp et al. (2023) &lt;doi:10.1093/g3journal/jkad118&gt;.">
<meta property="og:image" content="https://lcrawlab.github.io/mvMAPIT/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mvMAPIT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/get_started.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    </li>
<li class="dropdown-header">Tutorials</li>
    <li>
      <a href="articles/tutorial-docker-mvmapit.html">Dockerized mvMAPIT</a>
    </li>
    <li>
      <a href="articles/tutorial-simulations.html">Simulate Traits</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Studies</li>
    <li>
      <a href="articles/study-compare-p-value-combine-methods.html">Empirical comparison of P-value combination methods in mvMAPIT</a>
    </li>
    <li>
      <a href="articles/study-phillips-bnabs.html">Synergistic epistasis in binding affinity landscapes</a>
    </li>
    <li>
      <a href="articles/study-wtccc-mice.html">Joint modeling of hematology traits yields epistatic signal in stock of mice</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lcrawlab/mvMAPIT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="multivariate-mapit-documentation-">Multivariate MAPIT Documentation <img src="reference/figures/logo.png" align="right" alt="" width="120"><a class="anchor" aria-label="anchor" href="#multivariate-mapit-documentation-"></a>
</h1></div>
<p><a href="https://github.com/lcrawlab/mvMAPIT/actions/workflows/check-standard.yaml" class="external-link"><img src="https://github.com/lcrawlab/mvMAPIT/actions/workflows/check-standard.yaml/badge.svg" alt="R CMD check"></a> <a href="https://github.com/lcrawlab/mvMAPIT/actions/workflows/docker-image.yml" class="external-link"><img src="https://github.com/lcrawlab/mvMAPIT/actions/workflows/docker-image.yml/badge.svg" alt="Docker Image CI"></a> <a href="https://cranlogs.r-pkg.org/badges/grand-total/mvMAPIT" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/mvMAPIT" alt="CRAN downloads"></a></p>
<p>Find the full package documentation including examples and articles here: <a href="https://lcrawlab.github.io/mvMAPIT/">Multivariate MAPIT Documentation</a>.</p>
<div class="section level2">
<h2 id="the-multivariate-marginal-epistasis-test-mvmapit">The multivariate MArginal ePIstasis Test (mvMAPIT)<a class="anchor" aria-label="anchor" href="#the-multivariate-marginal-epistasis-test-mvmapit"></a>
</h2>
<p>This R package is a generalization of the <a href="https://github.com/lorinanthony/MAPIT" class="external-link">MAPIT implementation</a> by Crawford et al. (2017)<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> for any number of traits as described by Stamp et al. (2023)<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<p>mvMAPIT is implemented as a set of R and C++ routines, which can be carried out within an R environment.</p>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Profiling phenotypic heterogenity in cellular populations has recieved a renewed interested recently due to the increased accuracy of single cell sequencing and the decrease in sequencing costs. As a result, new statistical methods have been developed to characterize the phenotypic heterogenity that result from single cell sequencing experiments such as single cell RNA-sequencing (scRNA-Seq) experiments. One popular way of carrying out this task is through clustering followed by performing additional statistical tests post-clustering to select salient genes that contribute to cluster identity. Current methods are limited in that the number of clusters in the data must be known prior to clustering and post-clustering variable selection does not leverage information about the cluster formation in order to elect salient genes.</p>
<p>We present a method for the Nonparametric CLUstering of SIngle cell PopulatiONs (NCLUSION) - a sparse Bayesian Nonparametric model based on a Hierarchical Dirichlet Process mixture model. By combining a sparsity-promoting spike-and-slab prior on the cluster means and a nonparametric Hierarchical Dirichlet Process prior on the number of clusters, NCLUSION jointly learns the salient genes defining cluster identity and an optimal number of clusters needed to partition the data. Together with reasonable model approximations using varaiational inference, our proposed approach is scalable to large single cell RNA sequencing experiments.</p>
</div>
<div class="section level3">
<h3 id="the-method">The Method<a class="anchor" aria-label="anchor" href="#the-method"></a>
</h3>
<p>The Nonparametric CLUstering of SIngle cell PopulatiONs (NCLUSION) is a sparse Bayesian Nonparametric model based on a Hierarchical Dirichlet Process mixture model which aims to identify single cell RNA sequencing clusters and the genes defining cluster identity. It does through the use of a sparsity-promoting spike-and-slab prior on the cluster means and a nonparametric Hierarchical Dirichlet Process prior on the number of clusters. The key idea behind the concept NCLUSION is the treatment of cluster formation as a shift in expression of particular genes away from a global mean. This mean shift is modeled by the spike-and-slab prior. The number of clusters is modeled via a stick-breaking proccess that allocates more clusters as the amount of variation in the data grows. The learning of the model parameters proceeds via a variational-EM algorithm that allows the method to scale with the number of cells in the data. As an overview of NCLUSION and its corresponding software implementation, we will assume that we have access to an expression matrix from a scRNA-seq study on N cells and G genes denoted as X where X is an N x G matrix of expression counts with G denoting the number of genes.</p>
<p>The goal of NCLUSION is to identify the phenotypic clusters in the data while jointly identifying the genes that define these clusters. To accomplish this, we asses the quality of clustering using both labels and label-free metrics. We also perform gene-pathway enrichment to asses the quality of elected genes.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The package needs compilation but the released version can be installed from CRAN.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"mvMAPIT"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="the-r-environment">The R Environment<a class="anchor" aria-label="anchor" href="#the-r-environment"></a>
</h3>
<p>R is a widely used, free, and open source software environment for statistical computing and graphics. The most recent version of R can be downloaded from the <a href="https://cran.r-project.org/" class="external-link">Comprehensive R Archive Network (CRAN)</a>. CRAN provides precompiled binary versions of R for Windows, macOS, and select Linux distributions that are likely sufficient for many users’ needs. Users can also install R from source code; however, this may require a significant amount of effort. For specific details on how to compile, install, and manage R and R-packages, refer to the manual <a href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html" class="external-link">R Installation and Administration</a>.</p>
</div>
<div class="section level3">
<h3 id="r-packages-required-for-mvmapit">R Packages Required for mvMAPIT<a class="anchor" aria-label="anchor" href="#r-packages-required-for-mvmapit"></a>
</h3>
<p>mvMAPIT requires the installation of the following R libraries:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/devtools/index.html" class="external-link"> devtools </a></li>
<li><a href="https://cran.r-project.org/web/packages/ggplot2/index.html" class="external-link"> ggplot2 </a></li>
<li><a href="https://cran.r-project.org/web/packages/reticulate/index.html" class="external-link"> reticulate </a></li>
<li><a href="https://cran.r-project.org/web/packages/R.utils/index.html" class="external-link"> R.utils </a></li>
<li><a href="https://cran.r-project.org/web/packages/stringr/index.html" class="external-link"> stringr </a></li>
<li><a href="https://cran.r-project.org/web/packages/Matrix/index.html" class="external-link"> Matrix </a></li>
<li><a href="https://cran.r-project.org/web/packages/Seurat/index.html" class="external-link"> Seurat </a></li>
<li><a href="https://cran.r-project.org/web/packages/dplyr/index.html" class="external-link"> dplyr </a></li>
<li><a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html" class="external-link"> RColorBrewer </a></li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link"> SingleCellExperiment </a></li>
<li><a href="https://cran.r-project.org/web/packages/tidyverse/index.html" class="external-link"> tidyverse </a></li>
<li><a href="https://cran.r-project.org/web/packages/scales/index.html" class="external-link"> scales </a></li>
<li><a href="https://cran.r-project.org/web/packages/cowplot/index.html" class="external-link"> cowplot </a></li>
<li><a href="https://cran.r-project.org/web/packages/RCurl/index.html" class="external-link"> RCurl </a></li>
<li><a href="https://cran.r-project.org/web/packages/optparse/index.html" class="external-link"> optparse </a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html" class="external-link"> ComplexHeatmap </a></li>
<li><a href="https://cran.r-project.org/web/packages/circlize/index.html" class="external-link"> circilize </a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html" class="external-link"> clusterProfiler </a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/enrichplot.html" class="external-link"> enrichplot </a></li>
<li><a href="https://cran.r-project.org/web/packages/geneset/index.html" class="external-link"> geneset </a></li>
<li><a href="https://cran.r-project.org/web/packages/genekitr/index.html" class="external-link"> genekitr </a></li>
<li><a href="https://cran.r-project.org/web/packages/patchwork/index.html" class="external-link"> patchwork </a></li>
<li><a href="https://www.bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html" class="external-link"> org.Hs.eg.db </a></li>
</ul>
<p>The easiest method to install these packages is with the following example command entered in an R shell:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">install.packages</span>(<span class="kw">c</span>( <span class="st">'checkmate'</span>, </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a>                    <span class="st">'CompQuadForm'</span>, </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a>                    <span class="st">'dplyr'</span>, </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a>                    <span class="st">'foreach'</span>, </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a>                    <span class="st">'harmonicmeanp'</span>, </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a>                    <span class="st">'logging'</span>, </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a>                    <span class="st">'mvtnorm'</span>, </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>                    <span class="st">'Rcpp'</span>, </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>                    <span class="st">'RcppAlgos'</span>, </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a>                    <span class="st">'RcppArmadillo'</span>, </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>                    <span class="st">'RcppParallel'</span>, </span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>                    <span class="st">'RcppSpdlog'</span>, </span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>                    <span class="st">'testthat'</span>, </span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>                    <span class="st">'tidyr'</span>), </span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a>                    <span class="dt">dependencies =</span> <span class="ot">TRUE</span>);</span></code></pre></div>
<p>Alternatively, one can also <a href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Installing-packages" class="external-link">install R packages from the command-line</a>.</p>
</div>
<div class="section level3">
<h3 id="installing-mvmapit-from-sources">Installing mvMAPIT from Sources<a class="anchor" aria-label="anchor" href="#installing-mvmapit-from-sources"></a>
</h3>
<p>The easiest way to install the package from sources is to change into the directory of mvMAPIT and run <code>R CMD INSTALL . --preclean</code>. The <code>--preclean</code> flag makes sure that the latest state is run.</p>
</div>
<div class="section level3">
<h3 id="c-functions-required-for-mapit">C++ Functions Required for MAPIT<a class="anchor" aria-label="anchor" href="#c-functions-required-for-mapit"></a>
</h3>
<p>The code in this repository assumes that basic Fortran and C++ libraries and compilers are already set up on the running personal computer or cluster. If not, the mvMAPIT functions and necessary Rcpp packages will not work properly. A simple option is to use <a href="https://gcc.gnu.org/" class="external-link">gcc</a>. macOS users may use this collection by installing the <a href="https://brew.sh/index.html" class="external-link">Homebrew package manager</a> and then typing the following into the terminal:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="ex">brew</span> install gcc</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="openmp">OpenMP<a class="anchor" aria-label="anchor" href="#openmp"></a>
</h3>
<p>Note that mvMAPIT takes advantage of <a href="https://www.openmp.org/" class="external-link">OpenMP</a>, an API for multi-platform shared-memory parallel programming in C/C++. This is to speed up the computational time of the modeling algorithm. Unfortunately, macOS does not currently support OpenMP under the default compiler. A work around to use OpenMP in R on macOS can be found <a href="https://thecoatlessprofessor.com/programming/openmp-in-r-on-os-x/" class="external-link">here</a>. mvMAPIT can be compiled without OpenMP, but we recommend using it if applicable for scalability.</p>
</div>
<div class="section level3">
<h3 id="known-issues">Known Issues<a class="anchor" aria-label="anchor" href="#known-issues"></a>
</h3>
<ul>
<li><p>When your compiler changes, some R package dependencies might need to be recompiled. This is likely the case if the compilation error explicitly names an R package in the local library.</p></li>
<li><p>On macOS, you might need to run <code>brew reinstall z3</code> to fix <code>'libz3.4.11.dylib' (no such file)</code> related errors (<a href="https://github.com/Homebrew/discussions/discussions/3920" class="external-link">clang issues</a>).</p></li>
<li><p>For extra tips on how to run C++ on macOS, please visit <a href="https://seananderson.ca/2013/11/18/rcpp-mavericks.html" class="external-link uri">https://seananderson.ca/2013/11/18/rcpp-mavericks.html</a>.</p></li>
<li><p>For tips on how to avoid errors dealing with <code>-lgfortran</code> or <code>-lquadmath</code>, please visit <a href="https://thecoatlessprofessor.com/programming/rcpp-rcpparmadillo-and-os-x-mavericks-lgfortran-and-lquadmath-error/" class="external-link uri">https://thecoatlessprofessor.com/programming/rcpp-rcpparmadillo-and-os-x-mavericks-lgfortran-and-lquadmath-error/</a>.</p></li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="questions-and-feedback">Questions and Feedback<a class="anchor" aria-label="anchor" href="#questions-and-feedback"></a>
</h2>
<p>For questions or concerns with the MAPIT functions, please contact <a href="mailto:lcrawford@microsoft.com">Lorin Crawford</a> or <a href="mailto:julian_stamp@brown.edu">Julian Stamp</a>.</p>
<p>We appreciate any feedback you may have with our repository and instructions.</p>
</div>
<div class="section level2"><h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2></div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>L. Crawford, P. Zeng, S. Mukherjee, X. Zhou (2017). Detecting epistasis with the marginal epistasis test in genetic mapping studies of quantitative traits. <em>PLoS Genet</em>. <strong>13</strong>(7): e1006869. <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006869" class="external-link uri">https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006869</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>J. Stamp, A. DenAdel, D. Weinreich, L. Crawford (2023). Leveraging the Genetic Correlation between Traits Improves the Detection of Epistasis in Genome-wide Association Studies. <em>G3 Genes|Genomes|Genetics</em>, <strong>13</strong>(8), jkad118. doi: <a href="https://doi.org/10.1093/g3journal/jkad118" class="external-link uri">https://doi.org/10.1093/g3journal/jkad118</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>X. Zhou (2017). A unified framework for variance component estimation with summary statistics in genome-wide association studies. <em>Ann Appl Stat</em>. <strong>11</strong>(4): 2027-2051. <a href="https://projecteuclid.org/journals/annals-of-applied-statistics/volume-11/issue-4/A-unified-framework-for-variance-component-estimation-with-summary-statistics/10.1214/17-AOAS1052.full" class="external-link uri">https://projecteuclid.org/journals/annals-of-applied-statistics/volume-11/issue-4/A-unified-framework-for-variance-component-estimation-with-summary-statistics/10.1214/17-AOAS1052.full</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=mvMAPIT" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/lcrawlab/mvMAPIT/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mvMAPIT</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Julian Stamp <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-3014-6249" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lorin Crawford <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-0178-8242" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Julian Stamp, Lorin Crawford.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>