<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
</head>
<body data-spy="scroll" data-target="#toc">
    

  <div class="container template-article">
    <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
<div class="container">
  <div class="navbar-header">
    <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
      <span class="sr-only">Toggle navigation</span>
      <span class="icon-bar"></span>
      <span class="icon-bar"></span>
      <span class="icon-bar"></span>
    </button>
    <span class="navbar-brand">
      <a class="navbar-link" href="../index.html">NCLUSION</a>
      <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title=""></span>
    </span>
  </div>

  <div id="navbar" class="navbar-collapse collapse">
    <ul class="nav navbar-nav">
<li>
<a href="../articles/get_started.html">Get started</a>
</li>
<li>
<a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
  Tutorials
   
  <span class="caret"></span>
</a>
<ul class="dropdown-menu" role="menu">
<li class="divider">
  </li>
  <li>
    <a href="../tutorials/tissue_tutorial.html">Run NCLUSION on Dominguez
    Conde et al 2022 </a>
  </li>
  <li>
    <a href="../tutorials/pbmc_tutorial.html">Run NCLUSION on Zheng et al 2017</a>
  </li>
</ul>
  <li class="divider">
  </li>


<ul class="nav navbar-nav navbar-right">
<li>
<a href="https://github.com/microsoft/nclusion/" class="external-link">
  <span class="fab fa-github fa-lg"></span>
   
</a>
</li>
    </ul>
</div>
<!--/.nav-collapse -->
</div>
<!--/.container -->
</div>
<!--/.navbar -->
      

      </header><script src="study-compare-p-value-combine-methods_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Clustering Zheng et al 2017</h1>
    </div>
    <p>
    This tutorial demonstrates how to replicate NCLUSION clustering results for
    ~94k pure peripheral blood mononuclear cells obtained from 
    Zheng et al 2017 <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Open a bash
    terminal and follow along with the tutorial from your working directory. </p> 
    <h2 data-toc-skip>Required Tutorial Installations</h2>
    <p>First, install NCLUSION into a Julia environment following the
    installation instructions <a
    href="../index.html">here</a>. The NCLUSION tutorials also require
    installation of additional R and python libraries. These packages are NOT required for running NCLUSION only.</p>
      <h3 data-toc-skip>The R environment</h3>
    <p>R is a software environment for
    statistical computing and graphics. The most recent version of R can be
    downloaded from the <a href="https://cran.r-project.org/"
    class="external-link">Comprehensive R Archive Network (CRAN)</a>. CRAN
    provides precompiled binary versions of R for Windows, macOS, and select Linux
    distributions. For detailed
   installation instructions for R and R libraries click <a
    href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html"
    class="external-link">here</a>.</p>
    <p>The following R libraries are required for replicating results in the
    NCLUSION tutorials: </p>
  <ul>
  <li><a href="https://cran.r-project.org/web/packages/devtools/index.html" class="external-link"> devtools </a></li>
  <li><a href="https://cran.r-project.org/web/packages/reticulate/index.html" class="external-link"> reticulate </a></li>
  <li><a href="https://cran.r-project.org/web/packages/ggplot2/index.html" class="external-link"> ggplot2 </a></li>
  <li><a href="https://cran.r-project.org/web/packages/R.utils/index.html" class="external-link"> R.utils </a></li>
  <li><a href="https://cran.r-project.org/web/packages/stringr/index.html" class="external-link"> stringr </a></li>
  <li><a href="https://cran.r-project.org/web/packages/Matrix/index.html" class="external-link"> Matrix </a></li>
  <li><a href="https://cran.r-project.org/web/packages/Seurat/index.html" class="external-link"> Seurat </a></li>
  <li><a href="https://cran.r-project.org/web/packages/dplyr/index.html" class="external-link"> dplyr </a></li>
  <li><a href="https://cran.r-project.org/web/packages/RColorBrewer/index.html" class="external-link"> RColorBrewer </a></li>
  <li><a href="https://www.bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link"> SingleCellExperiment </a></li>
  <li><a href="https://cran.r-project.org/web/packages/tidyverse/index.html" class="external-link"> tidyverse </a></li>
  <li><a href="https://cran.r-project.org/web/packages/scales/index.html" class="external-link"> scales </a></li>
  <li><a href="https://cran.r-project.org/web/packages/cowplot/index.html" class="external-link"> cowplot </a></li>
  <li><a href="https://cran.r-project.org/web/packages/RCurl/index.html" class="external-link"> RCurl </a></li>
  <li><a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html" class="external-link"> ComplexHeatmap </a></li>
  <li><a href="https://cran.r-project.org/web/packages/circlize/index.html" class="external-link"> circilize </a></li>
  <li><a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html" class="external-link"> clusterProfiler </a></li>
  <li><a href="https://bioconductor.org/packages/release/bioc/html/enrichplot.html" class="external-link"> enrichplot </a></li>
  <li><a href="https://cran.r-project.org/web/packages/patchwork/index.html" class="external-link"> patchwork </a></li>
  </ul>
  <p>The easiest method to install these packages is with the following example
  command entered in an R shell:</p> <div class="sourceCode" id="cb2"><pre
  class="sourceCode R"><code class="sourceCode r">if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  
BiocManager::install(c("SingleCellExperiment", "ComplexHeatmap",
"clusterProfiler", "enrichplot"))

<span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">install.packages</span>(<span class="kw">c</span>(<span class="st">'patchwork'</span>, </span>
  <span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a>                    <span class="st">'ggplot2'</span>, </span>
  <span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a>                    <span class="st">'R.utils'</span>, </span>
  <span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a>                    <span class="st">'stringr'</span>, </span>
  <span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a>                    <span class="st">'Matrix'</span>, </span>
  <span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a>                    <span class="st">'Seurat'</span>, </span>
  <span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>                    <span class="st">'dplyr'</span>, </span>
  <span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>                    <span class="st">'RColorBrewer'</span>, </span>
  <span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>                    <span class="st">'tidyverse'</span>, </span>
  <span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>                    <span class="st">'scales'</span>, </span>
  <span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>                    <span class="st">'cowplot'</span>, </span>
  <span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>                    <span class="st">'RCurl'</span>,</span>
  <span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>                    <span class="st">'circilize'</span>, </span>
  <span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>                    <span class="st">'reticulate'</span>, </span>
  <span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>                    <span class="st">'devtools'</span>), </span> 
  <span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a>                    <span class="dt">dependencies =</span> <span class="ot">TRUE</span>);</span></code></pre></div>
  <p>You can find a table of the R package versions <a href="../tutorial_R_package_versions.csv">here.</a></p>
  <div class="section level3">
  <h3 id="installing-mvmapit-from-sources">The Python Environment<a class="anchor" aria-label="anchor" href="#installing-mvmapit-from-sources"></a>
  </h3>
  <p>We recommend that users install python packages into a new Anconda
  environment. Details for installing Anaconda and setting up a new environment
  can be found <a href="https://www.anaconda.com/download/" class="external-link">here</a>.</p>
  <p>Packages that are required to run NCLUSION tutorials include:</p>
  <ul>
  <li><a href="https://scanpy.readthedocs.io/en/stable/" class="external-link"> scanpy </a> </li>
  <li><a
  href="https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html"
  class="external-link">
  pandas </a> </li> 
  <li> <a href="https://matplotlib.org/stable/users/installing/index.html"
  class="external-link"> matplotlib </a> </li>
  <li> <a href="https://cluster-validity-indices.readthedocs.io/en/latest/guide.html#installation"
    class="external-link"> cvi </a> </li>
    <li> <a href="https://scikit-learn.org/stable/install.html"
      class="external-link"> scikit-learn </a> </li>
     <li><a href="https://www.gnu.org/software/wget/">wget</a> </li>
  <li> <a href="https://numpy.org/install/" class="external-link"> numpy</a>
  </li>
  <li> <a href="https://anndata.readthedocs.io/en/latest/" class="external-link"> anndata </a> </li>
  <li> <a href="https://gseapy.readthedocs.io/en/latest/" class="external-link"> gseapy </a> </li>
  </ul>
  <p>To install these packages, activate a new Anaconda environment and enter the
  following commands into a bash terminal:</p>
<pre><code>conda install scikit-learn pandas matplotlib numpy</pre></code>
  <pre><code>python -m pip install scanpy anndata gseapy cvi</code></pre>
  <pre><code>sudo apt-get install wget</code></pre>
  <p>Alternatively, you can download the environment with the exact versions by downloading this file: <a href="../environment.yml">environment.yml</a> and entering the following command into your bash terminal:</p>
  <pre><code>conda env create -f environment.yml</code></pre>
  </div>
  <h3 id="preprocess-data">Tutorial Scripts<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
  </h3> 
    Next, download the necessary tutorial scripts
    into your current working directory with the following bash commands.
    Alternatively, you can download the tutorial files as a zipped directory
    from <a
    href="https://microsoft.github.io/nclusion/tutorial_scripts.zip" download>here</a>, transfer them to your
    working directory, and unzip to use. 
    <pre><code>wget https://microsoft.github.io/nclusion/tutorial_scripts.zip
unzip tutorial_scripts.zip</code></pre>
    </p>
    <p><b>NOTE: to find more detailed documentation for each step of this tutorial, run each
      bash script with the
      <code>--help</code> flag.  </b></p>

<div class="section level3">
  <h2 id="preprocess-data">Download and Preprocess Data<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
  </h2>
  <p>The Zheng + 2017 data contains scRNA-seq data for ~94k pure
    peripheral blood mononuclear cells of 10 distinct cell types. The filtered, raw counts can be downloaded from <a
  href="https://www.10xgenomics.com/resources/datasets?query=cd&page=1&configure%5BhitsPerPage%5D=50&configure%5BmaxValuesPerFacet%5D=1000&refinementList%5BbiomaterialTypes%5D=&refinementList%5Bproduct.name%5D%5B0%5D=Single%20Cell%20Gene%20Expression&refinementList%5Bspecies%5D%5B0%5D=Human&refinementList%5BdiseaseStates%5D=&refinementList%5BanatomicalEntities%5D="
  class="external-link"e>here</a>.
  Alternatively, run the following <a href="https://www.gnu.org/software/wget/"
  class="external-link">wget</a> commands to create a directory named
  <code>tutorial_data/zheng2017/</code> in your current working directory and download the filtered count files for each
  cell type into that directory. </p>
  <pre><code>mkdir -p tutorial_data/
mkdir -p tutorial_data/zheng2017/
cd tutorial_data/zheng2017/

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/cytotoxic_t/cytotoxic_t_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/naive_cytotoxic/naive_cytotoxic_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd56_nk/cd56_nk_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd4_t_helper/cd4_t_helper_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/memory_t/memory_t_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/naive_t/naive_t_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/regulatory_t/regulatory_t_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd34/cd34_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/b_cells/b_cells_filtered_gene_bc_matrices.tar.gz

wget https://cf.10xgenomics.com/samples/cell-exp/1.1.0/cd14_monocytes/cd14_monocytes_filtered_gene_bc_matrices.tar.gz</code></pre>
  <p>Then, unzip the files in the data directory with the following bash command:</p>
  
  <pre><code>for f in `ls *.gz`; do
filename=$(basename -- "$f")
extension="${filename##*.}"
filename="${filename%.*}"
filename=$(basename -- "$filename")
filename="${filename%.*}"
echo "$filename";
mkdir "${filename}"
mv "${f}" "${filename}"
cd "${filename}"
tar -xzvf "$f"
cd ../
done 
cd ../../</code></pre>
  <p> Before running NCLUSION to cluster the Zheng + 2017 cells, the data must be
  converted to the AnnData type, and
  concatenated into a single data object. Then, to preprocess the data, we filter out genes that were expressed in less than 3 cells, cells
  with less than 200 genes expressed, greater than 20% mitochondrial reads, and
  less than 5% ribosomal reads to obtain 94,615 high-quality cells representing 10
  distinct cell types. We then subset the data to the top 5000 highly-variable
  genes. The data conversion, concatenation, and preprocessing can be
  executed by entering the following command into a bash terminal, which then
  saves the
  filtered, preprocessed data as an AnnData file located at
  <code>tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad</code>.</p> 
  <pre><code>. tutorial_scripts/preprocess.sh --datapath="tutorial_data/zheng2017/" --savepath="tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad" --data_name="zheng" --n_hvgs=5000</code></pre>
<div class="section level3">
<h2 id="run-nclusion">Run NCLUSION<a class="anchor" aria-label="anchor" href="#run-nclusion"></a>
</h2>
<p> To run NCLUSION from your working direcorty, enter the following bash
command.</p><p> <b>NOTE: set the <code>--julia_env</code> variable to the
relative path to your Julia project environment where NCLUSION is installed. </b> </p>
<pre><code>mkdir -p zheng2017_outputs

. tutorial_scripts/run_nclusion.sh --datapath='tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad' --julia_env='PATH/TO/JULIA/ENV' --data_name='zheng2017' --output_dir='zheng2017_outputs/'</code></pre>
<p>All tutorial output files can be found in the specified output
directory:
<code>zheng2017_outputs/</code>. All output files generated by NCLUSION can be
found in:
<code>zheng2017_outputs/outputs/EXPERIMENT_nclusion_zheng2017/DATASET_zheng2017_5000HVGs-94615N/YEAR_MONTH_DATE_TIME/</code>
</p><p> <b>NOTE: 'YEAR-MONTH-DATE-TIME'
labels in the output path/files corresponds to the time stamp of when the
results were generated. The time stamp is added automatically by NCLUSION. </b></p>
<p>In the output directory, you can find the following output files:</p>
<ol><li><b>zheng2017_5000HVGs-94615N_nclusion-YEAR-MONTH-DATE-TIME.csv</b>
  <ul><li>.csv file that contains the NCLUSION clustering assignments, where each row displays relevant metadata and
    clustering results for each cell. </li>
    <li> The 'condition' column
      displays the cell's experimental condition labels (all cells in this study
      have the same condition label).</li> 
      <li> The 'cell_id' column displays the cell's unique identifier label. </li>
    <li> The 'cell_type' column displays the cell's manual cell-type annotation
    published in this study. This column would be empty for unannotated datasets.
    </li>
  <li> The 'called_label' column displays the numerical equivalent of the
  'cell_type' label, mapped automatically by NCLUSION. </li>
  <li>The 'inferred_label' column displays the cell's NCLUSION cluster assignment.
  </li></ul></li>
  <li><b>5000G-YEAR-MONTH-DATE-TIME-pips.csv</b>
    <ul><li>.csv file that contains the Posterior Inclusion Probabilities (PIPs) of
      genes across clusters identified by NCLUSION. PIPs are used as a summary of
      evidence for a gene being associated with driving the identity of any
      phenotypic cluster. For a particular cluster, a higher PIP
      indicates that a gene is more significant to the formation of that cluster.
  </li> 
  </ul></li>
  <li><b>_QuickSummary_YEAR-MONTH-DATE-TIME.txt</b>
    <ul><li>.txt file that contains the input parameters that were used to run
    NCLUSION.</li></ul></li>
  <li><b>output.jld2</b>
  <ul><li>.jld2 file that contains all model parameters</li></ul></li>
  <li><b>YEAR-MONTH-DATE-TIME-Nk.csv</b><ul><li>.csv file that contains estimated
  number of cells in each cluster identified by NCLUSION</li></ul></li>
  </ol>
  </p>
<div class="section level3"><h2 id="eval-metrics">Analysis of NCLUSION
Clustering and Variable Selection<a class="anchor" aria-label="anchor" href="#eval-metrics"></a>
</h2></div>
<p><b>NOTE: In the following analyses, clusters are named according to the cluster mapping file:
  <code>tutorial_scripts/cluster_mapping/zheng_mapping.csv</code> which maps
  the NCLUSION labels to labels in [1:number of clusters], for aesthetic
  purposes. This mapping retains NCLUSION cluster memberships. </b></p> 
<h3 id="eval-metrics">A. Quantitative Evaluation Metrics<a class="anchor" aria-label="anchor" href="#eval-metrics"></a>
</h3>
<p> To evaluate the NCLUSION clustering results, the following command calculates
various evaluation metrics: Jaccard Indexs, Adjusted Rand Index (ARI) (extrinsic), McIntosh Evenness, and Gini-Simpson Index (intrinsic).
Extrinsic metric calculations are dependent on
experimental annotations of the data, that act as the "true" labels, while intrinsic metrics do not. Therefore, extrinsic metrics are
only calculated when "true" labels are provided in the "cell_type" observation
layer of the input AnnData object. The metric calculations are
automatically saved to <code>.csv</code> files in the specified output
directory. Each file is labeled according to the following format: '{name_of_metric}_NCLUSION_pbmc_5000hvgs.csv' </p>
<p>To calculate evaluation metrics, enter the following command into your bash terminal. </p> 
<p><b>NOTE: be sure to replace "/PATH/TO/NCLUSION/CLUSTERING/FILE" inputs with the
  relative path to your NCLUSION clustering results and "/PATH/TO/JULIA/ENV" with the path to your Julia environment where NCLUSION is installed. </b> </p>
<pre><code>python tutorial_scripts/utility_scripts/make_intermediates.py --input_path "/PATH/TO/NCLUSION/CLUSTERING/FILE" --title "NCLUSION Zheng et al 2017" --output_path_file "zheng2017_outputs/" --output_path_figure "zheng2017_outputs/" --output_suffix '' --dataset_ordering_mode "pbmc" --col_sum_threshold '0.15'

julia --project='PATH/TO/JULIA/ENV' tutorial_scripts/utility_scripts/get_Julia_jaccard_metrics.jl "zheng2017_outputs/pbmc_NCLUSION_mapped_clusters.csv" '4' '6' '1' 'NCLUSION' '5000' "pbmc" "" 'jaccardtrans' "zheng2017_outputs/"

python tutorial_scripts/utility_scripts/make_evalmetrics.py --labels_path "zheng2017_outputs/pbmc_NCLUSION_mapped_clusters.csv" --data_path 'tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad' --eval_metric "ginisimpsonindex" --condition_idx '1' --data_partition '' --dataset "pbmc" --method 'NCLUSION' --hvgs '5000' --output_path_file "zheng2017_outputs/" --output_suffix ''

python tutorial_scripts/utility_scripts/make_evalmetrics.py --labels_path "zheng2017_outputs/pbmc_NCLUSION_mapped_clusters.csv" --data_path 'tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad' --eval_metric "mcintosheveness" --condition_idx '1' --data_partition '' --dataset "pbmc" --method 'NCLUSION' --hvgs '5000' --output_path_file "zheng2017_outputs/" --output_suffix ''
</code></pre>
<h3 id="eval-metrics">B. Generate Cell Distribution Heatmap</h3><a class="anchor" aria-label="anchor" href="#eval-metrics"></a>
</h3>
<p>Next, generate a heatmap that displays how cells from each cell-type
(determined by the
"true" labels obtained from Zheng et al 2017) are distributed across
clusters that are identified by NCLUSION. Each row, <i>i</i>, represents one cell-type, and
each column, <i>j</i>, represents one cluster. Therefore the color intensity at
location <i>(i, j)</i> on the heatmap corresponds to percent of cells with
"true" cell-type label <i>i</i> that were assigned to cluster <i>j</i> by
NCLUSION.</p> <p>To generate the heatmap from the NCLUSION result, enter the
following commands into a bash terminal. </p> <p><b>NOTE: be sure to replace --nclusion_results input with the
  relative path to your NCLUSION clustering results.</b></p>
  <p><b>NOTE: if you want to load packages from a specific R library, append <code>--rlib="/PATH/TO/R/LIBRARY"</code> onto
    this command. </b> </p>
<pre><code>. tutorial_scripts/make_heatmap.sh --nclusion_results="/PATH/TO/NCLUSION/CLUSTERING/FILE" --pct_table="zheng2017_outputs/cell_distribution_table.csv" --cluster_dict="tutorial_scripts/cluster_mapping/zheng_mapping.csv" --save_heatmap="zheng2017_outputs/cell_distribution_heatmap.pdf" --data_name="zheng"</code></pre>
<p>This will produce a pdf file named <code>zheng2017_outputs/cell_distriubtion_heatmap.pdf</code>containing the cell distriubtion heatmap shown
below. The
column labels do not correspond to NCLUSION's cluster labels. The script labels
each cluster with their positional values in the heatmap for aesthetic purposes. A
mapping of the original NCLUSION cluster labels to these new positional labels can
be found in a file named <code>tutorial_scripts/cluster_mapping/zheng_mapping.csv</code></p>
<img src="../images/zheng/cell_distrib_zheng-1.png" width="800" height="400">
<h3 id="tsnes">C. Annotated TSNEs <a
  class="anchor" aria-label="anchor" href="#pip-heatmaps"></a>  
  </h3>
  <p>Now, we will plot the t-distributed stochastic neighbor embedding (tSNE)
  for the data annotated with various labels: the called cell type labels that
  were published along with the data, and the NCLUSION-inferred cell type labels. We
  will also generate tSNEs corresponding to the boolean label of each cells'
  membership to a given NCLUSION-inferred cluster, such that cells assigned to cluster
  <i>n</i> are colored and all other cells are greyed out. The
  total number of tSNEs produced are equal to the total number of
  NCLUSION-inferred clusters plus two. Run the following command to generate
  these plots. </p><p><b>NOTE: be sure to replace --nclusion_results input with the
  relative path to your NCLUSION clustering results. </b> </p>
  <pre><code>. tutorial_scripts/make_tsnes.sh --datapath="tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad" --nclusion_results="/PATH/TO/NCLUSION/CLUSTERING/FILE" --data_name="zheng2017"

mv figures/ zheng2017_outputs/</code></pre>
  <p>All tsnes produces by this script can be found in the output directory:
    <code>zheng2017_outputs/figures/</code>. The tSNEs colored by called
    cell-type, inferred cell-type, and Cluster 1 membership are shown below.</p>
    <img src="../images/zheng/tsne_zheng2017_called_cell_type.png" width="500"
    height="300"> 
    <img src="../images/zheng/tsne_zheng2017_inferred_label.png" width="400"
    height="300">
    <img src="../images/zheng/tsne_zheng2017_cluster_1.png" width="400"
    height="300"> 
  <h3 id="pip-heatmaps">D. Variable Selection Analysis <a
class="anchor" aria-label="anchor" href="#pip-heatmaps"></a> 
</h3>
<p>Next, analyze results of NCLUSION's variable selection functionality to gain
insight into the biological mechanisms driving the formation of each cluster.
The reulting plots are described in detail in the following sub-sections. To
generate these plots, enter the following command into a bash terminal. </p> 
<p><b>NOTE: if you want to load packages from a specific R library, append <code>--rlib="/PATH/TO/R/LIBRARY"</code> onto
this command.  </b></p>
<p><b>NOTE: be sure to replace --nclusion_results, --pips, and --path_to_nk inputs with the
relative path to your clustering assignment, PIP, and Nk results files (generated by
NCLUSION). </b> </p>
<pre><code>. tutorial_scripts/analyze_results.sh --datapath="tutorial_data/zheng2017/5000hvgs_zheng2017_preprocessed.h5ad" --pips="/PATH/TO/PIPS/FILE" --nclusion_results="/PATH/TO/NCLUSION/CLUSTERING/FILE" --output_dir="zheng2017_outputs/" --data_name='zheng' --mapping="tutorial_scripts/cluster_mapping/zheng_mapping.csv" --path_to_nk="PATH/TO/NK/FILE"</code></pre>
<p>All plots produced by this script can be found in the specified output
directory: <code>zheng2017_outputs/</code>
  directory. Each plot is described in detail below. </p>
<h4 id="pip-heatmaps">i. Adjusted Posterior Inclusion Probability (PIP) Heatmap
<a
  class="anchor" aria-label="anchor" href="#pip-heatmaps"></a> 
  </h4>
  <p>We generated a heatmap that displays adjusted Posterior Inclusion
    Probabilities (PIPs) of genes across
    clusters identified by NCLUSION. We use PIPs as a summary of evidence for gene being associated with driving the identity of any phenotypic cluster.
    For a particular cluster, a
    higher PIP indicates that a gene is more significant to the formation of that
    cluster. Before generating this heatmap, the plotting script adjusted the PIP
    values to account for
    the number of clusters in which it is deemed significant. For example, if a gene
    is deemed to be significant in driving all clusters, its PIP score will be
    scaled down, as it is less helpful for identifying important differences between each
    cluster compared to unique significant genes that are identified in fewer clusters. Only
  genes with adjusted PIPs above a threshold of 0.5 are included in this analysis.
  The plot is saved to <code>zheng2017_outputs/heatmap_AdjustedPips_KbyG_defaultclustering.pdf</code>.
  </p> 
<img src="../images/zheng/adjusted_pips_zheng-1.png" height="800" width="800">
<h4 id="effectsize-heatmaps">ii. Effect Size Heatmap
  <a
    class="anchor" aria-label="anchor" href="#effectsize-heatmaps"></a> 
    </h4>
    <p>The script also generated heatmap representing the effect size of the genes included
      in the adjusted PIP heatmap, which is shown below. For a given cluster, the effect size
    indicates the direction of expression (+ or -) for each gene in that cluster.
    The plot is saved to <code>zheng2017_outputs/heatmap_EffectSizeSignMat_KbyG_basedonAdjPips.pdf</code></p>
<img src="../images/zheng/effect_size_zheng-1.png" height="800" width="800"> 
<h4 id="violin-plot">iii. Module Expression Violin Analysis<a
    class="anchor" aria-label="anchor" href="#violin-plot"></a> 
    </h4>
    <p>In addition to the heatmaps, this script generated module expression violin
      plots for each cluster. Each violin plot compares the expression of
      significant genes identified in a particular cluster to the expression of
      that group of genes in the other clusters. The module expression analysis for
    Cluster 1 is shown below. The plots are saved to
    <code>zheng2017_outputs/violinplot_tsne_normedmodulescore_colored_nolegend_Cluster_NGenes.pdf</code>
    where N is equal to the cluster label. </p>
      <img src="../images/zheng/violin_cluster1-1.png" width="600"
      height="400">  
  <h4 id="go-enrich">v. Gene-set Over-Enrichment (GO) Analysis<a
    class="anchor" aria-label="anchor" href="#go-enrich"></a> 
    </h4>
    <p>Lastly, this script produced biological pathway
      enrichment plots. For each cluster, this analysis uses a
      hypergeometric test to calculate the enrichment of genes in a supplied gene
      set (genes with significant PIPs [>0.5] in that cluster). The GO analysis
    for Cluster 1 is shown below. The plots are saved to
    <code>zheng2017_outputs/GO_bp_enrichmentplots_Cluster_NGenes.pdf</code>
    where N is equal to the cluster label.</p>
      <img src="../images/zheng/GO_cluster1-1.png" width="400"
      height="800"> 
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p> Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan
  Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory,
  Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura,
  Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander
  Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb,
  William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen,
  Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of
  single cells. Nature Communications, 8(11):14049, Jan 2017. ISSN 2041-1723. doi: 10.1038/
  ncomms14049 <a href="https://doi.org/10.1038/ncomms14049" class="external-link uri">https://doi.org/10.1038/ncomms14049</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Chibuikem Nwizu, Lorin Crawford, Ava Amini, and Madeline Hughes</p>
</div>

      </footer>
</div>

<style>pre{
  position: relative;
  margin: 5px 0 ;
  padding: 1.75rem 0 1.75rem 1rem;

  /* more stuff */
}

.page-header{
  margin: 50px 0 0;
}

pre button{
  position: absolute;
  top: 5px;
  right: 5px;

  /* more stuff */
} 
pre {  
  word-wrap: normal;  
}  
</style>
<script>const copyButtonLabel = "Copy";

  // use a class selector if available
  let blocks = document.querySelectorAll("pre");
  
  blocks.forEach((block) => {
    // only add button if browser supports Clipboard API
    if (navigator.clipboard) {
      let button = document.createElement("button");
  
      button.innerText = copyButtonLabel;
      block.appendChild(button);
  
      button.addEventListener("click", async () => {
        await copyCode(block);
      });
    }
  });
  
  async function copyCode(block) {
    let code = block.querySelector("code");
    let text = code.innerText;
  
    await navigator.clipboard.writeText(text);
  }</script>


  

  </body>
</html>