<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Reference</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fisher's combine method on mvmapit return â€” fishers_combined"><meta property="og:description" content="This function takes in the p-values tibble that mvmapit returned. It then
computes the combined p-values grouped by variant id."><meta property="og:image" content="https://lcrawlab.github.io/mvMAPIT/logo.png"><meta name="google-site-verification" content="7EMmOVQ0-Feh--5i1ogdQqVYz7Azip5Je-UeBYXP6fA"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

  <div class="container template-article">
    <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
<div class="container">
  <div class="navbar-header">
    <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
      <span class="sr-only">Toggle navigation</span>
      <span class="icon-bar"></span>
      <span class="icon-bar"></span>
      <span class="icon-bar"></span>
    </button>
    <span class="navbar-brand">
      <a class="navbar-link" href="../index.html">NCLUSION</a>
      <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title=""></span>
    </span>
  </div>

  <div id="navbar" class="navbar-collapse collapse">
    <ul class="nav navbar-nav">
<li>
<a href="../articles/get_started.html">Get started</a>
</li>
<li>
<a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
  Tutorials
   
  <span class="caret"></span>
</a>
<ul class="dropdown-menu" role="menu">
<li class="divider">
  </li>
<li><a href="../tutorials/galenAML_tutorial.html">Run
NCLUSION on van Galen
et al 2019</a></li>
  <li>
    <a href="../tutorials/tissue_tutorial.html">Run NCLUSION on Dominguez
    Conde et al 2022 </a>
  </li>
  <li>
    <a href="../tutorials/pdac_tutorial.html">Run NCLUSION on Raghavan et al 2021</a>
  </li>
  <li>
    <a href="../tutorials/pbmc_tutorial.html">Run NCLUSION on Zheng et al 2017</a>
  </li>
</ul>
  <li class="divider">
  </li>


<ul class="nav navbar-nav navbar-right">
<li>
<a href="https://github.com/microsoft/nclusion/" class="external-link">
  <span class="fab fa-github fa-lg"></span>
   
</a>
</li>
    </ul>
</div>
<!--/.nav-collapse -->
</div>
<!--/.container -->
</div>
<!--/.navbar -->

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>run_nclusion()</h1>
    </div>

    <div class="ref-description">
    <p>This function runs the NCLUSION method</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span
    class="fu">run_nclusion</span><span class="op">(</span><span
    class="va">datafilename1,</span><span class="va"> KMax,</span><span
    class="va"> alpha1,</span><span class="va"> gamma1,</span><span class="va">
    seed,</span><span class="va"> elbo_ep,</span><span class="va">
    dataset,</span><span class="va">
    outdir,</span><span class="va"> logger=nothing,</span><span class="va">
    num_iter=500,</span><span class="va"> save_metrics=false</span><span
    class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>datafilename1</dt>
<dd><p>Absolute path to preprocessed AnnData object with scRNA-seq expression
values (cells x genes). The AnnData.obs_names layer must be occupied with
the unique cell IDs, the AnnData.var_names with unique gene identifiers,
AnnData.X with the expression matrix, and AnnData.obs["condition"] with the
condition label for each cell. <i>Optional:</i> if the dataset has been
pre-annotated, AnnData.obs["cell_type"] may be occupied with these annotations.
For more detailed information on the required input data structure see <a
href="../articles/get_started.html">Get started</a>.
</p></dd>


<dt>KMax</dt>
<dd><p>Maximum number of clusters NCLUSION will be initialized with</p></dd>


<dt>alpha1</dt>
<dd><p>second level concentration parameter (smaller values give fewer cluster)</p></dd>


<dt>gamma1</dt>
<dd><p>top level concentration parameter (smaller values give fewer cluster)</p></dd>


<dt>seed</dt>
<dd><p>Random seed</p></dd>


<dt>elbo_ep</dt>
<dd><p>Minumum tolerance in the change in elbo values between iterations. This
determines whether or not convergence is reached.</p></dd>


<dt>dataset</dt>
<dd><p>Name of the data set being analyzed (for output naming purposes)</p></dd>


<dt>outdir</dt>
<dd><p>Absolute path to where NCLUSION's output will be written</p></dd>


<dt>logger</dt>
<dd><p>Logging object that writes algorithms progress to stdout [Default: nothing]</p></dd>


<dt>num_iter</dt>
<dd><p>Maximum number of iterations the variational inference algorithm will continue to run before stopping if convergence is not reached [Default: 500]</p></dd>



<dt>save_metrics</dt>
<dd><p>Boolean that indicates whether or not NCLUSION should calculate
label-dependent clustering metrics (Note: Requires reference labels to be
provided in the AnnData.obs["cell_type"] of the input data object)
[Default: false]</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>

<p>Dictionary of all estimated parameters, including PIPs (yjk_) and cluster propbability (rtik_)</p>
    </div>
    <div id="saved-files">
      <h2>Saved Output Files</h2>
  
  <p>The following output files are automatically saved to
  <b>outputs/EXPERIMENT_nclusion_{DATA-NAME}/DATASET_{DATA-NAME}_{NUM-GENES}HVGs-{NUM-CELLS}N/{YEAR_MONTH_DATE_TIME}/ </b>
  which is automatically generated in the specified output
  directory, <code>outdir</code>, after running NCLUSION. </p> <p><b>NOTE: {DATA-NAME} corresponds to the <code>dataset</code> input variable.
    {NUM-GENES} corresponds to the number of genes included in the input AnnData
  object. {NUM-CELLS} corresponds to the number of cells included in the input
  AnnData object. {YEAR-MONTH-DATE-TIME} corresponds to the time-stamp at which the
    outputs were generated (calculated automatically by NCLUSION). </b> 
  <ol><li><b>{DATA-NAME}_{NUM-GENES}HVGs-{NUM-CELLS}N_nclusion-{YEAR-MONTH-DATE-TIME}.csv</b>
    <ul><li>.csv file that contains the NCLUSION clustering assignments, where each row displays relevant metadata and
    clustering results for each cell. </li>
    <li> The 'condition' column
      displays the cell's experimental condition labels (all cells in this study
      have the same condition label).</li> 
      <li> The 'cell_id' column displays the cell's unique identifier label. </li>
  <li>The 'inferred_label' column displays the cell's NCLUSION cluster assignment.
  </li></ul> <b>NOTE:</b> the following columns are only filled if alternative cell-type
  annotations are provided in the <code>AnnData.obs['cell_type']</code> layer of
the input data. Otherwise they remain empty:
<ul> <li> The 'cell_type' column displays the cell's alternative cell-type
annotation obtained from a different study.
  </li>
<li> The 'called_label' column displays the numerical equivalent of the
'cell_type' label, mapped automatically by NCLUSION. </li></ul></li>
  <li><b>{NUM-GENES}G-{YEAR-MONTH-DATE-TIME}-pips.csv</b>
    <ul><li>.csv file that contains the Posterior Inclusion Probabilities (PIPs) of
      genes across clusters identified by NCLUSION. PIPs are used as a summary of
      evidence for a gene being associated with driving the identity of any
      phenotypic cluster. For a particular cluster, a higher PIP
      indicates that a gene is more significant to the formation of that cluster.
  </li> 
  </ul></li>
  <li><b>_QuickSummary_{YEAR-MONTH-DATE-TIME}.txt</b>
    <ul><li>.txt file that contains the input parameters that were used to run
    NCLUSION.</li></ul></li>
  <li><b>output.jld2</b>
  <ul><li>.jld2 file that contains all model parameters</li></ul></li>
  <li><b>{YEAR-MONTH-DATE-TIME}-Nk.csv</b><ul><li>.csv file that contains estimated
  number of cells in each cluster identified by NCLUSION</li></ul></li>
  <li><b>*_Summary.csv</b></li>
  <ul><li><b>NOTE: </b>These files are ONLY generated if alternative cell-type annotations are provided in the
    <code>AnnData.obs['cell_type']</code> layer and the
    <code>save_metrics</code> input variable is set to <code>true</code>.</li>
    <li>There should be .csv files following this labeling format, corresponding to
  the quantitative metrics calculated to evaluate NCLUSION performance on this
  dataset. </li>
  <li>The files are named according to: {METRIC}_Summary.csv where
  {METRIC} corresponds to the name of the metric that was calculated. The
  following metrics are calculated automatically by NCLUSION: Adjusted Rand Index (ARI), Jaccard Index, Normalized
  Mutual Information (NMI), V-Measure, and Variation of Information (VarInfo).
  All metric calculations dependent on the <code>AnnData.obs['cell_type']</code> annotations.</li>
</ul></ol></p>
      </div>



      </footer></div>

  

      <style>pre{
        position: relative;
        margin: 5px 0 ;
        padding: 1.75rem 0 1.75rem 1rem;
      
        /* more stuff */
      }
      
      pre button{
        position: absolute;
        top: 5px;
        right: 5px;
      
        /* more stuff */
      } 
      pre {  
        word-wrap: normal;  
      }  
      </style>
      <script>const copyButtonLabel = "Copy";
      
        // use a class selector if available
        let blocks = document.querySelectorAll("pre");
        
        blocks.forEach((block) => {
          // only add button if browser supports Clipboard API
          if (navigator.clipboard) {
            let button = document.createElement("button");
        
            button.innerText = copyButtonLabel;
            block.appendChild(button);
        
            button.addEventListener("click", async () => {
              await copyCode(block);
            });
          }
        });
        
        async function copyCode(block) {
          let code = block.querySelector("code");
          let text = code.innerText;
        
          await navigator.clipboard.writeText(text);
        }</script>

  

  </body></html>